### DISCUSSION TITLE
Recognition of long and multiple key presses

### DISCUSSION THREAD

**Discussion by UweSchwo** [elgatosf/elgato#9] [Ideas]
I'm coding button functionalities since meanwhile 19 years (using AHK and extra keys on my keyboard). The recognition of long button presses or multiple button presses (up to 6 times) was among the first things I implemented. 

On my 2 Stream Decks I'm already using "long presses" to get edit mode concerning the current button function. So in my Office-Profiles the buttons raise VBA-macros but when I long press the button I get the VBA editor with buttons macro in focus.  

The best implementation of such a feature seems to have multiple edit registers in the dashboard for "long press" or maybe "double press".
Another way for multiple keypress recognition could be a more flexible key-folder functionality: Pressing a button opens a new folder, this already works. But in addition there's an option to automatically raise a button function and return to the previous profile view after some milliseconds without an additional press. So with nested folders we have a multiple button press recognition.

**Comment by ChekTek** [comment]
Hi UweSchwo,

If I understand correctly, you are looking for a "quick folder" like action where the key opens the folder for a set duration or a secondary action to be fired and then returns to the profile. This is a cool suggestion, I will provide this feedback to the team.

**Reply by UweSchwo** [reply]
Thanks, this seems to be the easiest way for multiple key press recognition. I try to describe how to use folders for this:

**1.** User has active "Profil1" and press a key that opens a folder "FolderA" with the new "timed / triggered return" feature set. 

**2.** The user see the "FolderA" keys, this may be only one at the place the previously pressed key was located.

**Variant 3. 1:**  The user press the key again within saying 500 ms. Thus the predefined function of that folder key will be launched.  Than automatically "Profil1" will be reactivated (may be after some key animation or just after some 700 ms so the user can see the pressed folder key for a while).

**Variant 3.2:** The user press any other key of "FolderA" within the 500 ms time limit.  Keys could be organized around the original pressed key. Thus the predefined function of that other folder key will be launched.  Than automatically "Profil1" will be reactivated.

**Variant 3.3:** The user doesn't press any key the next 500 ms. 
**Variant 3.3.1**: If the "FolderA" layout contains a key labeled as default action (for instance  a "*" inside the key title) this key action will be launched. Than automatically "Profil1" will be reactivated.
**Variant 3.3.1**: If the "FolderA" layout has no default key "Profil1" will be reactivated without any action.

If this would work with nested folders we have not only double press recognition but multiple key recognition.

Use case (in "Streaming Background", cause this seems to be important): A user could scale the micro level to "1", "2" or "3" value depending on how often he press the key. While he press the key multiple he see the current level numbers on the key.

Concerning the implementation of "Long presses" there's an plugin called "BarRaider SuperMacro" that already helps me to do this.

**Reply by ChekTek** [reply]
There are some other plugins in the store that support long presses for added functionality. We give developers the option to implement a "Long press" and "Multiple key presses" by keeping the `keyDown` and `keyUp` events separate. Developers tend not to implement these things due to the added complexity of keeping track of the time a key is down etc. Perhaps in the future, I can create some content around how to implement these things so that plugin developers are more aware of what is possible.

**Reply by UweSchwo** [reply]
Thanks for the response.
I do understand the problem. Honestly the time and trigger management are the reasons why I didn't yet implemented this for myself (with the SD AHK implementation). But having this as a core feature would make many things much easier. 
From users point of view I see no problems, he would see some new easy to use core features. So when I used BarRaiders long press implementation I at first didn't even remark that there were some ms-settings.

BTW: I've a list of 15 other ideas to further improve the Stream Decks but don't like to overcharge this way of communication. Finally every thing has to be separately observed.

**Comment by ChekTek** [comment]
Feel free to share more improvement ideas as new discussions! We love hearing new ideas for Stream Deck.

**Reply by UweSchwo** [reply]
Thanks. I'm not quite sure where to post which problem. 
Here seems to be the place for improvements concerning the SDK. 
In difference to earlier Elgato announcements this seems not the right place for ideas concerning the core software itself (like the one above). 
Following this I should post most of my ideas under https://e.lga.to/FeatureRequest.
My 10 hardware suggestions (from no glare buttons to keyboard integration) I transmitted via support e-mail.

**Reply by ChekTek** [reply]
You are correct! This discussion board is focused on SDK related topics. For general feature requests [https://e.lga.to/FeatureRequest](https://e.lga.to/FeatureRequest) is great!
